homeassistant/components/mqtt/update.py

```diff
@@ -141,25 +137,104 @@ def _setup_from_config(self, config: ConfigType) -> None:
         }
 
+    @callback
+    def _handle_state_message_received(self, msg: ReceiveMessage) -> None:
+        """Handle receiving state message via MQTT."""
+        payload = self._templates[CONF_VALUE_TEMPLATE](msg.payload)
+
+        if not payload or payload == PAYLOAD_EMPTY_JSON:
+            _LOGGER.debug(
+                "Ignoring empty payload '%s' after rendering for topic %s",
+                payload,
+                msg.topic,
+            )
+            return
+
+        json_payload: _MqttUpdatePayloadType = {}
+        try:
+            rendered_json_payload = json_loads(payload)
+            if isinstance(rendered_json_payload, dict):
+                _LOGGER.debug(
+                    (
+                        "JSON payload detected after processing payload '%s' on"
+                        " topic %s"
+                    ),
+                    rendered_json_payload,
+                    msg.topic,
+                )
+                json_payload = cast(_MqttUpdatePayloadType, rendered_json_payload)
+            else:
+                _LOGGER.debug(
+                    (
+                        "Non-dictionary JSON payload detected after processing"
+                        " payload '%s' on topic %s"
+                    ),
+                    payload,
+                    msg.topic,
+                )
+                json_payload = {"installed_version": str(payload)}
+        except JSON_DECODE_EXCEPTIONS:
+            _LOGGER.debug(
+                (
+                    "No valid (JSON) payload detected after processing payload '%s'"
+                    " on topic %s"
+                ),
+                payload,
+                msg.topic,
+            )
+            json_payload["installed_version"] = str(payload)
+
+        if "installed_version" in json_payload:
+            self._attr_installed_version = json_payload["installed_version"]
+
+        if "latest_version" in json_payload:
+            self._attr_latest_version = json_payload["latest_version"]
+
+        if "title" in json_payload:
+            self._attr_title = json_payload["title"]
+
+        if "release_summary" in json_payload:
+            self._attr_release_summary = json_payload["release_summary"]
+
+        if "release_url" in json_payload:
+            self._attr_release_url = json_payload["release_url"]
+
+        if "entity_picture" in json_payload:
+            self._entity_picture = json_payload["entity_picture"]
+
+    @callback
+    def _handle_latest_version_received(self, msg: ReceiveMessage) -> None:
+        """Handle receiving latest version via MQTT."""
+        latest_version = self._templates[CONF_LATEST_VERSION_TEMPLATE](msg.payload)
+
+        if isinstance(latest_version, str) and latest_version != "":
+            self._attr_latest_version = latest_version
+
     def _prepare_subscribe_topics(self) -> None:
         """(Re)Subscribe to topics."""
         topics: dict[str, Any] = {}
 
         def add_subscription(
```

Comment:
Same comment as fan. Would be nice to have `add_subscription` be a bound method instead of an inner function